drivers/regulator/regulator_npm1300.c

```diff
@@ -0,0 +1,613 @@
+	default:
+		return -ENODEV;
+	}
+}
+
+static int regulator_npm1300_set_buck_mode(const struct device *dev, uint8_t chan,
+					   regulator_mode_t mode)
+{
+	switch (mode) {
+	case NPM1300_BUCK_MODE_PWM:
+		return reg_write(dev, BUCK_BASE, offsetof(struct buck_reg_t, pwm[chan].set), 1U);
+	case NPM1300_BUCK_MODE_AUTO:
+		return reg_write(dev, BUCK_BASE, offsetof(struct buck_reg_t, pwm[chan].clr), 1U);
+	default:
+		return -ENOTSUP;
+	}
+}
+
+static int regulator_npm1300_set_ldsw_mode(const struct device *dev, uint8_t chan,
+					   regulator_mode_t mode)
+{
+	switch (mode) {
+	case NPM1300_LDSW_MODE_LDO:
+		return reg_write(dev, LDSW_BASE, offsetof(struct ldsw_reg_t, ldosel[chan]), 1U);
+	case NPM1300_LDSW_MODE_LDSW:
+		return reg_write(dev, LDSW_BASE, offsetof(struct ldsw_reg_t, ldosel[chan]), 0U);
+	default:
+		return -ENOTSUP;
+	}
+}
+
+int regulator_npm1300_set_mode(const struct device *dev, regulator_mode_t mode)
+{
+	const struct regulator_npm1300_config *config = dev->config;
+
+	switch (config->source) {
+	case NPM1300_SOURCE_BUCK1:
+		return regulator_npm1300_set_buck_mode(dev, 0, mode);
+	case NPM1300_SOURCE_BUCK2:
+		return regulator_npm1300_set_buck_mode(dev, 1, mode);
+	case NPM1300_SOURCE_LDO1:
+		return regulator_npm1300_set_ldsw_mode(dev, 0, mode);
+	case NPM1300_SOURCE_LDO2:
+		return regulator_npm1300_set_ldsw_mode(dev, 1, mode);
+	default:
+		return -ENODEV;
+	}
+}
+
+int regulator_npm1300_get_mode(const struct device *dev, regulator_mode_t *mode)
+{
+	const struct regulator_npm1300_config *config = dev->config;
+
+	switch (config->source) {
+	case NPM1300_SOURCE_BUCK1:
+	case NPM1300_SOURCE_BUCK2:
+	case NPM1300_SOURCE_LDO1:
+	case NPM1300_SOURCE_LDO2:
+		return -ENOTSUP;
+	default:
+		return -ENODEV;
+	}
+}
+
+int regulator_npm1300_enable(const struct device *dev)
+{
+	const struct regulator_npm1300_config *config = dev->config;
+
+	switch (config->source) {
+	case NPM1300_SOURCE_BUCK1:
+		return reg_write(dev, BUCK_BASE, offsetof(struct buck_reg_t, en[0].set), 1U);
+	case NPM1300_SOURCE_BUCK2:
+		return reg_write(dev, BUCK_BASE, offsetof(struct buck_reg_t, en[1].set), 1U);
+	case NPM1300_SOURCE_LDO1:
+		return reg_write(dev, LDSW_BASE, offsetof(struct ldsw_reg_t, en[0].set), 1U);
+	case NPM1300_SOURCE_LDO2:
+		return reg_write(dev, LDSW_BASE, offsetof(struct ldsw_reg_t, en[1].set), 1U);
+	default:
+		return -ENODEV;
+	}
+}
+
+int regulator_npm1300_disable(const struct device *dev)
+{
+	const struct regulator_npm1300_config *config = dev->config;
+
+	switch (config->source) {
+	case NPM1300_SOURCE_BUCK1:
+		return reg_write(dev, BUCK_BASE, offsetof(struct buck_reg_t, en[0].clr), 1U);
+	case NPM1300_SOURCE_BUCK2:
+		return reg_write(dev, BUCK_BASE, offsetof(struct buck_reg_t, en[1].clr), 1U);
+	case NPM1300_SOURCE_LDO1:
+		return reg_write(dev, LDSW_BASE, offsetof(struct ldsw_reg_t, en[0].clr), 1U);
+	case NPM1300_SOURCE_LDO2:
+		return reg_write(dev, LDSW_BASE, offsetof(struct ldsw_reg_t, en[1].clr), 1U);
```

Comment:
not a fan of using offsetof, I think it's better to use definitions for the offsets, I'd say it's more idiomatic in Zephyr.